File: runme/mcgen/local/datan/datsrc/auxgrd.f
      SUBROUTINE AUXGRD(X,GRD,N,NRED,LIST,FUNC)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(DELTA=1.D-11,CUT=1.D-9)
      DIMENSION X(*),GRD(*),LIST(*)
      EXTERNAL FUNC
      IL=0
      IF(NRED.GT.0) THEN
        DO 10 I=1,N
          IF(LIST(I).EQ. 1.OR. NRED.EQ.N) THEN
            IL=IL+1
            ARG=ABS(X(I))
            IF(ARG.LT.CUT) ARG=CUT
            DEL=DELTA*ARG
            SAV=X(I)
            X(I)=SAV+DEL
            FP=FUNC(X,N)
            X(I)=SAV-DEL
            FM=FUNC(X,N)
            X(I)=SAV
            GRD(IL)=(FP-FM)/(DEL+DEL)
          END IF
   10   CONTINUE
      END IF
      END

File: runme/mcgen/local/datan/datsrc/auxgrd.c
#include "datsrc.h"

void LIBFUNC auxgrd_(double P_T *x, double P_T *grd, integer const P_T *n,
             integer const P_T *nred, integer const P_T *list,
             double (CBFUNC P_T *func)(double const P_T *,
             integer const P_T *))
{
    /* System generated locals */
    integer i__1;
    double d__1;

    /* Local variables */
    integer i, il;
    double fm, fp, del, arg, sav;

    /* Parameter adjustments */
    --list;
    --grd;
    --x;

    /* Function Body */
    il = 0;
    if (*nred > 0) {
        i__1 = *n;
        for (i = 1; i <= i__1; ++i) {
            if (list[i] == 1 || *nred == *n) {
                ++il;
                arg = (d__1 = x[i], abs(d__1));
                if (arg < 1e-9) {
                    arg = 1e-9;
                }
                del = arg * 1e-11;
                sav = x[i];
                x[i] = sav + del;
                fp = (*func)(&x[1], n);
                x[i] = sav - del;
                fm = (*func)(&x[1], n);
                x[i] = sav;
                grd[il] = (fp - fm) / (del + del);
            }
        }
    }
    return;
} /* auxgrd_ */

